### PR TITLE
#49: Support sanitizing label names in PrometheusMetricsExporter, introduce PrometheusMetricsExporterBuilder

### PR DESCRIPTION
**Pull request by devromik**

### CODE REVIEW

File: metrics-facade-spring/metrics-facade-prometheus-spring-boot-autoconfigure/src/main/java/com/ringcentral/platform/metrics/spring/prometheus/MfPrometheusAutoConfiguration.java [hunk 1]
```diff
@@ -71,15 +72,16 @@ public PrometheusMetricsExporter prometheusMetricsExporter(
 
         MfPrometheusConfig config = configBuilder.build();
 
-        InstanceSamplesProvider<? extends PrometheusSample, ? extends PrometheusInstanceSample> instanceSamplesProvider =
+        final var instanceSamplesProvider =
```
Comment (mairovichaa) [inline metrics-facade-spring/metrics-facade-prometheus-spring-boot-autoconfigure/src/main/java/com/ringcentral/platform/metrics/spring/prometheus/MfPrometheusAutoConfiguration.java:75]: It's pretty difficult to understand from the context type of created instance.

I would suggest to declare in both places types explicitly.

**Review by mairovichaa — APPROVED** [review]

File: metrics-facade-spring/metrics-facade-prometheus-spring-boot-autoconfigure/src/main/java/com/ringcentral/platform/metrics/spring/prometheus/MfPrometheusAutoConfiguration.java (hunk 1 above)
Comment (devromik) [reply metrics-facade-spring/metrics-facade-prometheus-spring-boot-autoconfigure/src/main/java/com/ringcentral/platform/metrics/spring/prometheus/MfPrometheusAutoConfiguration.java:75]: ok, will roll it back, thank you!

File: metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/PrometheusMetricsExporter.java [hunk 2]
```diff
@@ -42,109 +39,38 @@ public enum Format {
...
     public static final Format DEFAULT_FORMAT = PROMETHEUS_TEXT_O_O_4;
     public static final String NAME_PARTS_DELIMITER = "_";
-    public static final boolean DEFAULT_LOWER_CASE_NAME = false;
+    public static final boolean DEFAULT_CONVERT_NAME_TO_LOWER_CASE = false;
     public static final Locale DEFAULT_LOCALE = Locale.ENGLISH;
 
     private final Format defaultFormat;
-    private final InstanceSamplesProvider<? extends PrometheusSample, ? extends PrometheusInstanceSample> instanceSamplesProvider;
     private final boolean convertNameToLowercase;
     private final Locale locale;
```
Comment (ikurovsky) [inline metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/PrometheusMetricsExporter.java:47]: Do we need locale after sanitizing?

**Review by ikurovsky — APPROVED** [review]

File: metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/PrometheusMetricSanitizer.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    List<String> sanitizeLabelNames(List<String> labelNames);
+
+    /**
+     * Sanitizes a single Prometheus label name according to the data model specification.
+     *
+     * @param labelName the label name to sanitize
+     * @return the sanitized label name
+     */
+    String sanitizeLabelName(String labelName);
```
[the remark this answers is not in the document]
Comment (devromik) [reply metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/PrometheusMetricSanitizer.java:32]: Discussed offline that it wont be supported for now.

File: metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/PrometheusMetricsExporter.java (hunk 2 above)
Comment (devromik) [reply metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/PrometheusMetricsExporter.java:47]: ok, will remove param, thank you!

File: metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/DefaultPrometheusMetricSanitizer.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            labelNames;
+    }
+
+    @Override
+    public String sanitizeLabelName(String labelName) {
+        if (!sanitizeLabelName) {
+            return labelName;
+        }
+
+        String result = LABEL_NAME_FORBIDDEN_CHAR_PATTERN.matcher(labelName).replaceAll(NAME_PARTS_DELIMITER);
```
[the remark this answers is not in the document]
Comment (devromik) [reply metrics-facade-prometheus/src/main/java/com/ringcentral/platform/metrics/reporters/prometheus/DefaultPrometheusMetricSanitizer.java:65]: The next line resolves this issue by adding the prefix if needed: return isLetter(result.charAt(0)) ? result : nonLetterLabelNameFirstCharPrefix + result;

**Pull request merged**